pkg/operator/controller/controller.go

```diff
@@ -0,0 +1,864 @@
+			Type:   conditions.ConditionDegraded,
+			Status: corev1.ConditionFalse,
+		})
+	}
+
+	logger.Info("Finished degraded check", "conditions", cr.Status.Conditions)
+	return degraded, nil
+}
+
+func checkDeploymentReady(deployment *appsv1.Deployment) bool {
+	desiredReplicas := deployment.Spec.Replicas
+	if desiredReplicas == nil {
+		desiredReplicas = &[]int32{1}[0]
+	}
+
+	if *desiredReplicas != deployment.Status.Replicas ||
+		deployment.Status.Replicas != deployment.Status.ReadyReplicas {
+		return false
+	}
+
+	return true
+}
+
+func (r *ReconcileVMImportConfig) add(mgr manager.Manager) error {
+	// Create a new controller
+	c, err := controller.New("vm-import-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	r.controller = c
+
+	if err = r.watchVMImportConfig(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReconcileVMImportConfig) watchVMImportConfig() error {
+	return r.controller.Watch(&source.Kind{Type: &vmimportv1alpha1.VMImportConfig{}}, &handler.EnqueueRequestForObject{})
+}
+
+func (r *ReconcileVMImportConfig) watchDependantResources(cr *vmimportv1alpha1.VMImportConfig) error {
+	r.watchMutex.Lock()
+	defer r.watchMutex.Unlock()
+
+	if r.watching {
+		return nil
+	}
+
+	resources, err := r.getAllResources(cr)
+	if err != nil {
+		return err
+	}
+
+	if err = r.watchResourceTypes(resources); err != nil {
+		return err
+	}
+
+	r.watching = true
+
+	return nil
+}
+
+func (r *ReconcileVMImportConfig) getAllDeployments(cr *vmimportv1alpha1.VMImportConfig) ([]*appsv1.Deployment, error) {
+	var result []*appsv1.Deployment
+
+	resources, err := r.getAllResources(cr)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, resource := range resources {
+		if deployment, ok := resource.(*appsv1.Deployment); ok {
+			result = append(result, deployment)
+		}
+	}
+
+	return result, nil
+}
+
+func (r *ReconcileVMImportConfig) getOperatorArgs(cr *vmimportv1alpha1.VMImportConfig) *OperatorArgs {
+	result := *r.operatorArgs
+
+	if cr != nil {
+		if cr.Spec.ImagePullPolicy != "" {
+			result.PullPolicy = string(cr.Spec.ImagePullPolicy)
+		}
+	}
+
+	return &result
+}
+
+func (r *ReconcileVMImportConfig) getAllResources(cr *vmimportv1alpha1.VMImportConfig) ([]runtime.Object, error) {
+	var resources []runtime.Object
+
+	if deployClusterResources() {
+		rs := createCRDResources()
+		resources = append(resources, rs...)
+	}
+
+	nsrs := createControllerResources(r.getOperatorArgs(cr))
+	resources = append(resources, nsrs...)
+
+	return resources, nil
+}
+
+func createControllerResources(args *OperatorArgs) []runtime.Object {
+	return []runtime.Object{
+		resources.CreateServiceAccount(args.Namespace),
+		resources.CreateControllerRole(args.Namespace),
+		resources.CreateControllerRoleBinding(args.Namespace),
+		resources.CreateControllerDeployment("vm-import-controller", args.Namespace, args.ControllerImage, args.PullPolicy, int32(1)),
```

Comment:
I am using it for testing where I fake different object to make sure update works correctly.

Reply:
In that case, there will be a mismatch since the deployment is being created with name `vm-import-controller` but in method `isControllerDeployment` the deployment name is compared to `vm-import-deployment`
```
func isControllerDeployment(d *appsv1.Deployment) bool {
	return d.Name == "vm-import-deployment"
}
```